apps/emqx_gateway_nats/src/emqx_nats_frame.erl

```diff
@@ -0,0 +1,502 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2025 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%--------------------------------------------------------------------
+
+-module(emqx_nats_frame).
+
+-behaviour(emqx_gateway_frame).
+
+-include("emqx_nats.hrl").
+
+-export([
+    initial_parse_state/1,
+    serialize_opts/0,
+    serialize_pkt/2,
+    parse/2,
+    format/1,
+    type/1,
+    is_message/1
+]).
+
+-export([
+    message/1,
+    subject/1,
+    sid/1,
+    queue_group/1,
+    reply_to/1,
+    headers/1,
+    payload/1
+]).
+
+-define(INIT_STATE(S, B), #{state := init, buffer := B} = S).
+-define(ARGS_STATE(S, B), #{state := args, buffer := B} = S).
+-define(HEADERS_STATE(S, B), #{state := headers, buffer := B} = S).
+-define(PAYLOAD_STATE(S, B), #{state := payload, buffer := B} = S).
+
+-elvis([{elvis_style, dont_repeat_yourself, disable}]).
+
+%%--------------------------------------------------------------------
+%% API
+%%--------------------------------------------------------------------
+
+initial_parse_state(_) ->
+    #{
+        buffer => <<>>,
+        %% TODO: support to parse HPUB and HMSG
+        headers => false,
+        %% init | args | headers | payload
+        state => init,
+        %% in parsing frame, the current frame is stored in iframe
+        iframe => undefined
+    }.
+
+parse(Data0, ?INIT_STATE(State, Buffer)) ->
+    Data = <<Buffer/binary, Data0/binary>>,
+    parse_operation(Data, State);
+parse(Data0, ?ARGS_STATE(State, Buffer)) ->
+    Data = <<Buffer/binary, Data0/binary>>,
+    parse_args(Data, State#{buffer => <<>>});
+parse(Data0, ?HEADERS_STATE(State, Buffer)) ->
+    Data = <<Buffer/binary, Data0/binary>>,
+    parse_headers(Data, State#{buffer => <<>>});
+parse(Data0, ?PAYLOAD_STATE(State, Buffer)) ->
+    Data = <<Buffer/binary, Data0/binary>>,
+    parse_payload(Data, State#{buffer => <<>>}).
+
+%%--------------------------------------------------------------------
+%% Control frames
+
+parse_operation(<<"PING\r\n", Rest/binary>>, State) ->
+    return_ping(Rest, State);
+parse_operation(<<"ping\r\n", Rest/binary>>, State) ->
+    return_ping(Rest, State);
+parse_operation(<<"PONG\r\n", Rest/binary>>, State) ->
```

Comment:
does server send PING to clients ?

Reply:
Yes.

> Once a client establishes a connection to the NATS server, the server will continuously send PING messages to the client at a configurable interval.

https://docs.nats.io/reference/reference-protocols/nats-protocol#ping-pong